pkg/prometheus/asset_store.go

```diff
@@ -261,3 +287,30 @@ func (a *assetStore) getSecretKey(ctx context.Context, namespace string, sel v1.
 
 	return string(secret.Data[sel.Key]), nil
 }
+
+func verifyCert(caPEM, certPEM []byte) error {
+	roots := x509.NewCertPool()
+	ok := roots.AppendCertsFromPEM(caPEM)
+	if !ok {
+		return errors.New("failed to parse CA certificate")
+	}
+
+	block, _ := pem.Decode([]byte(certPEM))
+	if block == nil {
+		return errors.New("failed to parse certificate")
+	}
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse certificate")
+	}
+
+	opts := x509.VerifyOptions{
+		Roots: roots,
+	}
+
+	if _, err := cert.Verify(opts); err != nil {
```

Comment:
I'm not sure we should go that deep in the verification, especially since the user may ask to skip the TLS verification in the pod/service monitor definition. I would only check that the CA PEM is valid.

Reply:
I had forgotten to check if the TLS verification was skipped before calling this function, it should be good now.

If we don't discard unsigned certificates right away, the invalid configuration will be hand over to Prometheus which might discard it at some point. To me, this ends up being the same issue as with the invalid keys/certificates.

Reply:
Hmm, we shouldn't verify the client certificate/key against the CA: client/key are used for client authentication (e.g. let the server authenticate Prometheus) while the CA is used by Prometheus to verify the certificate presented by the server.

Reply:
Makes sense, thank you for the explanation.